[FEAT] 안드로이드 로그 디버그모드 관리

stopstone opened the pull request:
## 작업 내용
- 로그 확장을 통해 로깅 시스템 개선
- 안드로이드 기본 Log를 사용하면 릴리즈모드에서의 로그 관리가 어려워지는 것을 개선
- 디버그/릴리즈 환경별로 로그 레벨을 다르게 관리
- 로그 포맷팅, 태그 관리 필요
- 전역에서 일관된 로깅 시스템을 구축

## 확인 방법
utils/NearLog.kt에 로그 확삼 함수 구현하였습니다.

## 참고 사항

### 구현부
```kotlin
// Debug 로그
fun logd(
    message: String,
    tag: String = TAG,
) {
    if (!isLoggingEnabled()) return
    Log.d(tag, buildLogMessage(message))
}
```
디버그모드에서만 로그가 출력되도록 구현하였습니다

일반 Log.d()와는 다르게 tag를  message 파라미터 뒤에 구현하였습니다.
tag의 생략을 고려하여 위와 같이 구현하였습니다.


### 사용법
```kotlin
@HiltAndroidApp
class NearApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        logd("onCreate: ", "NearApplication")    // 1번째 - 커스텀 태그
        logd("onCreate: ")          // 2번째 - 기본 태그 "Near"
    }
}
```

### 결과
```
Near: [NearApplication::onCreate (NearApplication.kt:7)] onCreate: 
```

## 관련 이슈
- close #24

stopstone committed:
feat: 로그 유틸리티 추가

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/utils/logger/NearLog.kt:
```diff
@@ -0,0 +1,135 @@
+ * 스택 트레이스 추출에 실패하면 원본 메시지를 반환합니다
+ */
+private fun buildLogMessage(message: String): String {
+    return try {
```
꼼꼼하게 처리주셨네요👍
예외가 발생하면 어떤 상황에서 발생할까요? 만약 예외가 발생하고 catch되어 message로 나오는 상황이면 어떤 예외인지,어쩌면 예외가 발생했는지도 모르게 다른 부분을 디버깅 하는 상황이 생길 수 있을 것 같습니다!

stopstone replied:
코멘트 감사합니다 :D
우선 예외가 발생하는 경우는 로그에 해당 파일까지 남기는 형태로 작성하였는데, 파일을 찾는 방식이
앱의 스택 트레이스를 순회하면서 탐색하고 있습니다! 그래서 이 순회과정에서 문제가 생길 수 있어 try~catch로 예외처리를 두었어요

예외처리를 두긴 하였지만 웬만하면 에러가 발생 안하지 않을까.. 싶습니다

rhkrwngud445 reviewed (commented):
고생하셨습니다! 작업 감사해요👍
예외를 처리하는 관점에서 코멘트 남겼습니다! 예외를 회피할지, 핸들링할지의 지석님의 의견이 궁금합니다!

stopstone committed:
refactor: 로깅 시 호출자 정보 찾는 로직 개선

- 기존: 고정된 스택 인덱스를 사용하여 호출자 정보를 찾음
- 변경: 스택 트레이스를 순회하여 NearLog.kt 파일이 아닌 첫 번째 호출자를 찾도록 변경
- 추가: 스택 트레이스 접근 권한이 없거나 기타 예외 발생 시, 로그 메시지에 관련 정보를 포함하여 디버깅 용이성 향상

stopstone committed:
refactor: 로깅 유틸리티 `NearLog` 개선

- `NearLog`를 object로 변경하여 싱글턴으로 사용하도록 수정
- 로그 메시지 생성 로직을 `runCatching`으로 예외처리
- 중복되는 로그 출력 로직을 `writeLog` 함수로 통합
- `loge` 함수 오버로딩을 단일 함수로 통합하고, `name` 파라미터를 nullable로 변경
- 함수명을 `logv` -> `v`, `logd` -> `d` 등으로 축약하여 간결하게 변경
